File: leetcode_problems/typescript/src/lc367_valid_perfect_square.ts
function isPerfectSquareIsGood(value: number, target: number) {
  return value * value <= target;
}

function isPerfectSquare(num: number): boolean {
  let low = 0;
  let high = num + 1;
  while (high - low > 1) {
    let mid = low + Math.floor((high - low) / 2);
    if (isPerfectSquareIsGood(mid, num)) {
      low = mid;
    } else {
      high = mid;
    }
  }
  return low * low == num;
}

File: leetcode_problems/cpp/lc367_valid_perfect_square.cpp
class Solution
{
public:
    bool is_good(long value, long target)
    {
        return (value * value) <= target;
    }

    bool isPerfectSquare(long num)
    {
        long low = 0;
        long high = num + 1;
        while (high - low > 1)
        {
            long mid = low + (high - low) / 2;
            if (is_good(mid, num))
            {
                low = mid;
            }
            else
            {
                high = mid;
            }
        }
        return low * low == num;
    }
};